File: scripts/train_model.py
import os
# Force TensorFlow to use legacy Keras 2 instead of Keras 3
# This ensures TensorFlow.js compatibility
os.environ['TF_USE_LEGACY_KERAS'] = '1'

import numpy as np
import tensorflow as tf
from tensorflow.keras.preprocessing.image import load_img, img_to_array
import json

# Constants
BASE_DIR = os.path.dirname(os.path.abspath(__file__))
DATA_DIR = os.path.join(BASE_DIR, '../src/app/images')
OUTPUT_DIR = os.path.join(BASE_DIR, '../public/model')
DOC_DIR = os.path.join(DATA_DIR, 'document')
REGULAR_DIR = os.path.join(DATA_DIR, 'regular_iamge') # Preserving typo from JS source

IMAGE_SIZE = 224
NUM_CLASSES = 2
EPOCHS = 100
BATCH_SIZE = 4
LEARNING_RATE = 0.00005
MODEL_SAVE_PATH = os.path.join(BASE_DIR, 'saved_model.h5')

def load_images():
    images = []
    labels = []

    # Load document images (label 0)
    if os.path.exists(DOC_DIR):
        doc_files = [f for f in os.listdir(DOC_DIR) if f.lower().endswith('.jpg')]
        print(f"Found {len(doc_files)} document images")
        for file in doc_files:
            img_path = os.path.join(DOC_DIR, file)
            try:
                img = load_img(img_path, target_size=(IMAGE_SIZE, IMAGE_SIZE))
                img_array = img_to_array(img)
                img_array = img_array / 255.0
                images.append(img_array)
                labels.append(0)
            except Exception as e:
                print(f"Error loading {file}: {e}")

    # Load regular images (label 1)
    if os.path.exists(REGULAR_DIR):
        reg_files = [f for f in os.listdir(REGULAR_DIR) if f.lower().endswith('.jpg')]
        print(f"Found {len(reg_files)} regular images")
        for file in reg_files:
            img_path = os.path.join(REGULAR_DIR, file)
            try:
                img = load_img(img_path, target_size=(IMAGE_SIZE, IMAGE_SIZE))
                img_array = img_to_array(img)
                img_array = img_array / 255.0
                images.append(img_array)
                labels.append(1)
            except Exception as e:
                print(f"Error loading {file}: {e}")
    
    if not images:
        return np.array([]), np.array([])
        
    return np.array(images), tf.keras.utils.to_categorical(np.array(labels), NUM_CLASSES)

def train():
    # Ensure output directory exists
    if not os.path.exists(OUTPUT_DIR):
        os.makedirs(OUTPUT_DIR)
        
    print("Loading images...")
    X, y = load_images()
    
    if len(X) == 0:
        print("No images found. Exiting.")
        return

    print(f"Data loaded: {len(X)} images")
    
    # Load MobileNet v1 with alpha 0.25 using tf.keras.applications
    print("Loading MobileNet...")
    base_model = tf.keras.applications.MobileNet(
        input_shape=(IMAGE_SIZE, IMAGE_SIZE, 3),
        alpha=0.25,
        depth_multiplier=1,
        dropout=0.001,
        include_top=False,
        weights='imagenet',
        pooling=None
    )
    
    # Create new model using tf.keras
    inputs = tf.keras.Input(shape=(IMAGE_SIZE, IMAGE_SIZE, 3))
    # We need to make sure the base_model is running in inference mode
    x = base_model(inputs, training=False)
    x = tf.keras.layers.GlobalAveragePooling2D()(x)
    x = tf.keras.layers.Dense(100, activation='relu')(x)
    x = tf.keras.layers.Dropout(0.5)(x)
    outputs = tf.keras.layers.Dense(NUM_CLASSES, activation='softmax')(x)
    
    model = tf.keras.models.Model(inputs, outputs)
    
    # Freeze the MobileNet layers
    base_model.trainable = False
    
    model.compile(
        optimizer=tf.keras.optimizers.Adam(learning_rate=LEARNING_RATE),
        loss='categorical_crossentropy',
        metrics=['accuracy']
    )
    
    print("Training model...")
    model.fit(
        X, y,
        epochs=EPOCHS,
        batch_size=BATCH_SIZE,
        shuffle=True,
        validation_split=0.0
    )
    
    print(f"Saving Keras model to {MODEL_SAVE_PATH}...")
    model.save(MODEL_SAVE_PATH)
    
    # Save metadata.json (same as JS script)
    metadata = {
        'labels': ['document', 'regular']
    }
    
    metadata_path = os.path.join(OUTPUT_DIR, 'metadata.json')
    with open(metadata_path, 'w') as f:
        json.dump(metadata, f, indent=2)
    print(f"Metadata saved to {metadata_path}")

if __name__ == '__main__':
    train()

File: scripts/train-model.js
/* eslint-disable @typescript-eslint/no-require-imports */
const tf = require('@tensorflow/tfjs');
const { createCanvas, loadImage } = require('canvas');
const fs = require('fs');
const path = require('path');

const DATA_DIR = path.join(__dirname, '../src/app/images');
const OUTPUT_DIR = path.join(__dirname, '../public/model');
const DOC_DIR = path.join(DATA_DIR, 'document');
const REGULAR_DIR = path.join(DATA_DIR, 'regular_iamge');

const IMAGE_SIZE = 224;
const NUM_CLASSES = 2; // 0: document, 1: regular

// Custom IO Handler for Node.js filesystem saving
class NodeFileSystemSaveHandler {
  constructor(dir) {
    this.dir = dir;
  }

  async save(modelArtifacts) {
    if (!fs.existsSync(this.dir)) {
      fs.mkdirSync(this.dir, { recursive: true });
    }

    // Handle weightsManifest construction if missing
    let weightsManifest = modelArtifacts.weightsManifest;
    if (!weightsManifest && modelArtifacts.weightSpecs) {
       weightsManifest = [{
         paths: ['./weights.bin'],
         weights: modelArtifacts.weightSpecs
       }];
    }
    
    const modelJson = {
      modelTopology: modelArtifacts.modelTopology,
      format: modelArtifacts.format,
      generatedBy: modelArtifacts.generatedBy,
      convertedBy: modelArtifacts.convertedBy,
      weightsManifest: weightsManifest,
    };
    
    if (modelArtifacts.weightData != null) {
      // Save weights binary
      const weightsPath = path.join(this.dir, 'weights.bin');
      fs.writeFileSync(weightsPath, Buffer.from(modelArtifacts.weightData));
      
      // Update manifest paths to be relative to model.json
      if (modelJson.weightsManifest) {
        modelJson.weightsManifest.forEach(manifest => {
          manifest.paths = ['./weights.bin'];
        });
      }
    }

    fs.writeFileSync(
      path.join(this.dir, 'model.json'),
      JSON.stringify(modelJson, null, 2)
    );
    
    // Also save metadata.json for class names if we want
    const metadata = {
      labels: ['document', 'regular']
    };
    fs.writeFileSync(
      path.join(this.dir, 'metadata.json'),
      JSON.stringify(metadata, null, 2)
    );

    return {
      modelArtifactsInfo: {
        dateSaved: new Date(),
        modelTopologyType: 'JSON',
        modelTopologyBytes: JSON.stringify(modelArtifacts.modelTopology).length,
        weightSpecsBytes: 0,
        weightDataBytes: modelArtifacts.weightData ? modelArtifacts.weightData.byteLength : 0,
      }
    };
  }
}

async function loadImages() {
  const images = [];
  const labels = [];

  const docFiles = fs.readdirSync(DOC_DIR).filter(f => f.endsWith('.jpg'));
  const regFiles = fs.readdirSync(REGULAR_DIR).filter(f => f.endsWith('.jpg'));

  console.log(`Found ${docFiles.length} document images`);
  console.log(`Found ${regFiles.length} regular images`);

  for (const file of docFiles) {
    const imgPath = path.join(DOC_DIR, file);
    const img = await loadImage(imgPath);
    const canvas = createCanvas(img.width, img.height);
    const ctx = canvas.getContext('2d');
    ctx.drawImage(img, 0, 0);
    
    const tensor = tf.browser.fromPixels(canvas)
      .resizeNearestNeighbor([IMAGE_SIZE, IMAGE_SIZE])
      .toFloat()
      .div(tf.scalar(255.0))
      .expandDims();
    images.push(tensor);
    labels.push(0); // 0 for document
  }

  for (const file of regFiles) {
    const imgPath = path.join(REGULAR_DIR, file);
    const img = await loadImage(imgPath);
    const canvas = createCanvas(img.width, img.height);
    const ctx = canvas.getContext('2d');
    ctx.drawImage(img, 0, 0);
    
    const tensor = tf.browser.fromPixels(canvas)
      .resizeNearestNeighbor([IMAGE_SIZE, IMAGE_SIZE])
      .toFloat()
      .div(tf.scalar(255.0))
      .expandDims();
    images.push(tensor);
    labels.push(1); // 1 for regular
  }

  return {
    images: tf.concat(images),
    labels: tf.oneHot(tf.tensor1d(labels, 'int32'), NUM_CLASSES)
  };
}

async function train() {
  // Ensure CPU backend is used if others fail, but usually auto-detected.
  await tf.setBackend('cpu');
  console.log('Using backend:', tf.getBackend());

  const data = await loadImages();
  console.log('Data loaded');
  
  // Load MobileNet
  console.log('Loading MobileNet...');
  // Need to use global fetch or polyfill if not available, but Node 18+ has fetch.
  const mobilenet = await tf.loadLayersModel('https://storage.googleapis.com/tfjs-models/tfjs/mobilenet_v1_0.25_224/model.json');
  
  // Create a new model
  const layer = mobilenet.getLayer('conv_pw_13_relu');
  const truncatedMobileNet = tf.model({inputs: mobilenet.inputs, outputs: layer.output});
  
  const model = tf.sequential();
  model.add(truncatedMobileNet);
  model.add(tf.layers.globalAveragePooling2d({}));
  model.add(tf.layers.dense({units: 100, activation: 'relu'}));
  model.add(tf.layers.dropout({rate: 0.5})); // Add dropout to reduce overfitting
  model.add(tf.layers.dense({units: NUM_CLASSES, activation: 'softmax'}));

  // Freeze the MobileNet layers
  for (const layer of truncatedMobileNet.layers) {
    layer.trainable = false;
  }

  model.compile({
    optimizer: tf.train.adam(0.00005), // Lower learning rate for fine-tuning
    loss: 'categoricalCrossentropy',
    metrics: ['accuracy']
  });

  console.log('Training model...');
  await model.fit(data.images, data.labels, {
    epochs: 100, // More epochs
    batchSize: 4,
    validationSplit: 0.0,
    shuffle: true,
    callbacks: {
      onEpochEnd: (epoch, logs) => console.log(`Epoch ${epoch}: loss=${logs.loss.toFixed(4)}, acc=${logs.acc.toFixed(4)}`)
    }
  });

  console.log('Saving model...');
  // Use custom IO handler
  await model.save(new NodeFileSystemSaveHandler(OUTPUT_DIR));
  console.log('Model saved to ' + OUTPUT_DIR);
}

train().catch(console.error);
